File: python/verify.py
"""
Verifies that the parallel implementations produce the same set of branches
as the sequential implementations (after lexicographic sort by coordinates).
"""
import math
import sys
import numpy as np
from multiprocessing import Pool, freeze_support

from symmetric_sequential import generate_fractal_tree
from symmetric_parallel import _worker, _build_tasks
from asymmetric_sequential import generate_fractal_tree_asymmetric
from asymmetric_parallel import _worker as _worker_asym, _build_tasks as _build_tasks_asym


def _parallel_sym(trunk_length, ratio, branch_angle, min_length, num_processes):
    branch_angle_rad = math.radians(branch_angle)
    split_depth = max(1, math.ceil(math.log2(num_processes * 4)))
    start_angle = math.pi / 2
    end_x = trunk_length * math.cos(start_angle)
    end_y = trunk_length * math.sin(start_angle)

    upper = [(0.0, 0.0, end_x, end_y, 0)]
    new_len = trunk_length * ratio
    tasks  = _build_tasks(end_x, end_y, new_len, start_angle + branch_angle_rad,
                          ratio, branch_angle_rad, min_length, 1, split_depth, upper)
    tasks += _build_tasks(end_x, end_y, new_len, start_angle - branch_angle_rad,
                          ratio, branch_angle_rad, min_length, 1, split_depth, upper)
    with Pool(processes=num_processes) as pool:
        results = pool.map(_worker, tasks)
    return np.concatenate([np.array(upper, dtype=np.float64)] + results)


def _parallel_asym(trunk_length, left_ratio, right_ratio,
                   left_angle, right_angle, min_length, num_processes):
    left_rad  = math.radians(left_angle)
    right_rad = math.radians(right_angle)
    split_depth = max(1, math.ceil(math.log2(num_processes * 4)))
    start_angle = math.pi / 2
    end_x = trunk_length * math.cos(start_angle)
    end_y = trunk_length * math.sin(start_angle)

    upper = [(0.0, 0.0, end_x, end_y, 0)]
    tasks  = _build_tasks_asym(end_x, end_y, trunk_length * left_ratio,
                               start_angle + left_rad,
                               left_ratio, right_ratio, left_rad, right_rad,
                               min_length, 1, split_depth, upper)
    tasks += _build_tasks_asym(end_x, end_y, trunk_length * right_ratio,
                               start_angle - right_rad,
                               left_ratio, right_ratio, left_rad, right_rad,
                               min_length, 1, split_depth, upper)
    with Pool(processes=num_processes) as pool:
        results = pool.map(_worker_asym, tasks)
    return np.concatenate([np.array(upper, dtype=np.float64)] + results)


def _sort(branches):
    # Lexicographic sort: primary x1, then y1, x2, y2, depth
    idx = np.lexsort((branches[:, 4], branches[:, 3], branches[:, 2],
                      branches[:, 1], branches[:, 0]))
    return branches[idx]


def _check(name, seq, par):
    seq_s, par_s = _sort(seq), _sort(par)
    if seq_s.shape != par_s.shape:
        print(f"  FAIL  {name}  branch count: seq={len(seq_s)}, par={len(par_s)}")
        return False
    ok = np.allclose(seq_s, par_s, atol=1e-10, rtol=0)
    print(f"{'  OK  ' if ok else '  FAIL'}  {name}  ({len(seq_s)} branches)")
    return ok


def main():
    SYM  = dict(trunk_length=100.0, ratio=0.67, branch_angle=30.0, min_length=0.01)
    ASYM = dict(trunk_length=100.0, left_ratio=0.67, right_ratio=0.57,
                left_angle=35.0, right_angle=25.0, min_length=0.0023)

    print("=== Python correctness verification ===\n")
    all_ok = True

    seq_sym = generate_fractal_tree(
        0, 0, SYM['trunk_length'], math.pi / 2,
        SYM['ratio'], math.radians(SYM['branch_angle']), SYM['min_length'],
    )
    for n in [1, 2, 4, 8]:
        all_ok &= _check(f"symmetric   N={n}", seq_sym, _parallel_sym(**SYM, num_processes=n))

    seq_asym = generate_fractal_tree_asymmetric(
        0, 0, ASYM['trunk_length'], math.pi / 2,
        ASYM['left_ratio'], ASYM['right_ratio'],
        math.radians(ASYM['left_angle']), math.radians(ASYM['right_angle']),
        ASYM['min_length'],
    )
    for n in [1, 2, 4, 8]:
        all_ok &= _check(f"asymmetric  N={n}", seq_asym, _parallel_asym(**ASYM, num_processes=n))

    print()
    print("All checks passed." if all_ok else "One or more checks FAILED.")
    if not all_ok:
        sys.exit(1)


if __name__ == "__main__":
    freeze_support()
    main()

File: rust/src/verify.rs
use fractal_tree::{
    generate_fractal_tree, generate_fractal_tree_asymmetric,
    collect_tasks, collect_tasks_asymmetric,
    Branch,
};
use rayon::prelude::*;
use std::f64::consts::PI;
use std::error::Error;


fn sort_branches(branches: &mut Vec<Branch>) {
    // Round to 1e-6 precision before comparing so that the ~6e-15 floating-point
    // discrepancy in the symmetric trunk endpoint (hardcoded 0 vs. computed cos(PI/2)*L)
    // does not affect sort order.
    branches.sort_by_key(|b| {
        let r = |x: f64| -> i64 { (x * 1e6).round() as i64 };
        (r(b.x1), r(b.y1), r(b.x2), r(b.y2), b.depth)
    });
}

fn branches_equal(a: &[Branch], b: &[Branch], tol: f64) -> bool {
    a.len() == b.len()
        && a.iter().zip(b.iter()).all(|(x, y)| {
            (x.x1 - y.x1).abs() < tol
                && (x.y1 - y.y1).abs() < tol
                && (x.x2 - y.x2).abs() < tol
                && (x.y2 - y.y2).abs() < tol
                && x.depth == y.depth
        })
}

fn check(name: &str, mut seq: Vec<Branch>, mut par: Vec<Branch>) -> bool {
    sort_branches(&mut seq);
    sort_branches(&mut par);
    let ok = branches_equal(&seq, &par, 1e-10);
    println!(
        "{}  {}  ({} branches)",
        if ok { "  OK  " } else { "  FAIL" },
        name,
        seq.len()
    );
    ok
}

fn parallel_sym(
    trunk_length: f64,
    ratio: f64,
    branch_angle_rad: f64,
    min_length: f64,
    num_threads: usize,
) -> Vec<Branch> {
    let pool = rayon::ThreadPoolBuilder::new()
        .num_threads(num_threads)
        .build()
        .unwrap();
    let split_depth = (num_threads * 4).next_power_of_two().ilog2() as usize;

    let mut upper = vec![Branch { x1: 0.0, y1: 0.0, x2: 0.0, y2: trunk_length, depth: 0 }];
    let child_len = trunk_length * ratio;
    let mut tasks = collect_tasks(
        &mut upper, 0.0, trunk_length, child_len,
        PI / 2.0 + branch_angle_rad, ratio, branch_angle_rad, 1, split_depth,
    );
    tasks.extend(collect_tasks(
        &mut upper, 0.0, trunk_length, child_len,
        PI / 2.0 - branch_angle_rad, ratio, branch_angle_rad, 1, split_depth,
    ));

    let subtrees: Vec<Vec<Branch>> = pool.install(|| {
        tasks.par_iter()
            .map(|t| generate_fractal_tree(
                t.x, t.y, t.length, t.angle,
                ratio, branch_angle_rad, min_length, t.depth,
            ))
            .collect()
    });

    let mut all = upper;
    for sub in subtrees {
        all.extend(sub);
    }
    all
}

fn parallel_asym(
    trunk_length: f64,
    left_ratio: f64,
    right_ratio: f64,
    left_angle_rad: f64,
    right_angle_rad: f64,
    min_length: f64,
    num_threads: usize,
) -> Vec<Branch> {
    let pool = rayon::ThreadPoolBuilder::new()
        .num_threads(num_threads)
        .build()
        .unwrap();
    let split_depth = (num_threads * 4).next_power_of_two().ilog2() as usize;

    let mut upper = vec![Branch { x1: 0.0, y1: 0.0, x2: 0.0, y2: trunk_length, depth: 0 }];
    let mut tasks = collect_tasks_asymmetric(
        &mut upper, 0.0, trunk_length, trunk_length * left_ratio,
        PI / 2.0 + left_angle_rad,
        left_ratio, right_ratio, left_angle_rad, right_angle_rad, 1, split_depth,
    );
    tasks.extend(collect_tasks_asymmetric(
        &mut upper, 0.0, trunk_length, trunk_length * right_ratio,
        PI / 2.0 - right_angle_rad,
        left_ratio, right_ratio, left_angle_rad, right_angle_rad, 1, split_depth,
    ));

    let subtrees: Vec<Vec<Branch>> = pool.install(|| {
        tasks.par_iter()
            .map(|t| generate_fractal_tree_asymmetric(
                t.x, t.y, t.length, t.angle,
                left_ratio, right_ratio, left_angle_rad, right_angle_rad, min_length, t.depth,
            ))
            .collect()
    });

    let mut all = upper;
    for sub in subtrees {
        all.extend(sub);
    }
    all
}

fn main() -> Result<(), Box<dyn Error>> {
    println!("=== Rust correctness verification ===\n");

    let branch_angle_rad = 30_f64.to_radians();
    let left_angle_rad   = 35_f64.to_radians();
    let right_angle_rad  = 25_f64.to_radians();

    let seq_sym = generate_fractal_tree(
        0.0, 0.0, 100.0, PI / 2.0, 0.67, branch_angle_rad, 0.01, 0,
    );
    let seq_asym = generate_fractal_tree_asymmetric(
        0.0, 0.0, 100.0, PI / 2.0,
        0.67, 0.57, left_angle_rad, right_angle_rad, 0.0023, 0,
    );

    let mut all_ok = true;

    for n in [1_usize, 2, 4, 8] {
        let par = parallel_sym(100.0, 0.67, branch_angle_rad, 0.01, n);
        all_ok &= check(&format!("symmetric   N={n}"), seq_sym.clone(), par);
    }

    for n in [1_usize, 2, 4, 8] {
        let par = parallel_asym(100.0, 0.67, 0.57, left_angle_rad, right_angle_rad, 0.0023, n);
        all_ok &= check(&format!("asymmetric  N={n}"), seq_asym.clone(), par);
    }

    println!();
    if all_ok {
        println!("All checks passed.");
    } else {
        eprintln!("One or more checks FAILED.");
        std::process::exit(1);
    }

    Ok(())
}
